driver/src/main/java/org/neo4j/driver/internal/connector/socket/TrustOnFirstUseTrustManager.java

```diff
@@ -0,0 +1,167 @@
+/**
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.connector.socket;
+
+import java.io.BufferedReader;
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import javax.net.ssl.X509TrustManager;
+import javax.xml.bind.DatatypeConverter;
+
+/**
+ * References:
+ * http://stackoverflow.com/questions/6802421/how-to-compare-distinct-implementations-of-java-security-cert-x509certificate?answertab=votes#tab-top
+ */
+
+class TrustOnFirstUseTrustManager implements X509TrustManager
+{
+    /**
+     * A list of pairs (known_server_ip, certificate) are stored in this file.
+     * When establishing a SSL connection to a new server, we will save the server's IP and its certificate in this
+     * file.
+     * Then when we try to connect to a known server again, we will authenticate the server by checking if it provides
+     * the same certificate as the one saved in this file.
+     */
+    static final String KNOWN_CERTS_FILE_PATH = "known_certs";
```

Comment:
We should think some on where we put this file.. putting it directly in the working dir is probably the most sensible option, assuming we then provide a mechanism to change the location (eg. using Config or something). The alternative would be to put it in `System.getProperty("user.home") + "/.neo4j/known_certs"` or some such. 

I'd suggest that independent of where we put it, we should have a default name that makes it clear what it is. `neo4j_known_hosts` would be stronger in signalling that it's for Neo and that it's basically the same as the known_hosts file for SSH. To be further helpful, having a comment on the first line in the file to explain why it's there might be useful;

    # This file contains trusted certificates for Neo4j servers, it's created by Neo4j drivers. 
    # You can configure the location of this file by setting the NEO4J_TRUSTED_HOSTS_PATH environment var

Second, permissions for this file become important. I'm not sure how good java is at cross-platform permissions here, but if there's some way we can make the file readable/writeable only by the user that created the file, that'd be brilliant.